HPCC-19557 Spray partition calculation can't be aborted.

AttilaVamos opened the pull request:
Add abort handling to partitioner code

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

Tested manually with spray/despray regression cases and ECLWatch

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

AttilaVamos commented:
@jakesmith please review.

jakesmith commented on dali/ft/daftformat.cpp:
```diff
@@ -58,6 +60,29 @@ CPartitioner::CPartitioner()
+        lastAbortCheckTick = nowTick;
+    }
+    return aborting;
+}
```
This is duplicating the functionality (from filecopy.cpp).
It would be better to call back to the filesprayer's implementation.

You can implement a IAbortRequestCallback in the filespayer that you pass to setAbort, instead of the original one.

AttilaVamos replied:
Reworked

jakesmith commented:
@AttilaVamos - see inline comments. Worth avoiding duplicating the isAborting() functionality.

AttilaVamos commented on dali/ft/daftformat.cpp:
```diff
@@ -113,6 +138,8 @@ void CPartitioner::commonCalcPartitions()
     if (target)
         target->setOutput(0);
 
+    LocalAbortHandler localHandler(daftAbortHandler);
```
Removed

AttilaVamos commented:
@jakesmith updated, please re-review.

jakesmith commented on dali/ft/filecopy.hpp:
```diff
@@ -132,7 +132,7 @@ static const char * DfuOperatonStr[] =
         "DFUtransfer"
     };
-interface IFileSprayer : public IInterface
+interface IFileSprayer : public IInterface, IAbortRequestCallback
```
Isn't necessary to extend the IFileSprayer interface and doesn't really make sense for it to be a extension of IAbortRequestCallback, as it itself takes a IAbortRequestCallback.

Instead you can implement a IAbortRequestCallback inside FileSprayer and use it for the partitioner->setAbort calls, e.g.:
```
    class CAbortRequestCallback : implements IAbortRequestCallback
    {
        FileSprayer &sprayer;
    public:
        CAbortRequestCallback(FileSprayer &_sprayer) : sprayer(_sprayer) { }
        virtual bool abortRequested() { return sprayer.isAborting(); }
    } fileSprayerAbortChecker;
```

can make it a member class of FileSprayer

jakesmith commented on dali/ft/daftformat.cpp:
```diff
@@ -58,6 +58,22 @@ CPartitioner::CPartitioner()
+{
+    if (abortChecker && abortChecker->abortRequested())
+    {
+        LOG(MCdebugInfo, unknownJob, "CPartitioner::isAborting() -> Abort requested via callback");
```
There can be e.g. 400 partitioners ? 
FileSpray already logs the abort, so 400 extra lines is excessive. 

Not a lot of point in the 'aborting' flag either, as it's a quick call into the FileSprayer once aborted,
so might as well be just:
```
return abortChecker && abortChecker->abortRequested();
```

jakesmith reviewed (changes requested):
@AttilaVamos - a couple of comments.

AttilaVamos commented:
@jakesmith 2nd attempt, please review.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 5169f4f764e0e771d8eaaba6563122aeaaad1e80
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |88 |88 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |776 |776 |0 |
|test (thor) |707 |707 |0 |
|test (roxie) |820 |820 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   17 sec (00:00:17) |  190 sec (00:03:10) |   55 sec (00:00:55) |    6 sec (00:00:06) |   22 sec (00:00:22) | 1177 sec (00:19:37) |   17 sec (00:00:17) | 1484 sec (00:24:44) |

AttilaVamos committed:
HPCC-19557 Spray partition calculation can't be aborted.

Add abort handling to partitioner code

Tested manually with spray/despray regression cases and ECLWatch

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 9721eb5fe5b1fe80a9518b358eaff1830ed7907a
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |88 |88 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |776 |776 |0 |
|test (thor) |707 |707 |0 |
|test (roxie) |820 |820 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    6 sec (00:00:06) |  182 sec (00:03:02) |   56 sec (00:00:56) |    6 sec (00:00:06) |   22 sec (00:00:22) | 1185 sec (00:19:45) |   16 sec (00:00:16) | 1473 sec (00:24:33) |

AttilaVamos commented:
@jakesmith indentation is fixed, commits are squashed.

jakesmith commented:
@AttilaVamos - looks good.

@richardkchapman - ready to merge.
